src/saead/session.c

```diff
@@ -0,0 +1,324 @@
+    if (err)
+    {
+        return -EIO;
+    }
+
+    session_id[id_len] = '\0';
+
+    return sprintf(buf,
+                   "E0:%c:%s:C%c%c",
+                   id->initiator == POUCH_ROLE_DEVICE ? 'D' : 'S',
+                   session_id,
+                   algorithm == PSA_ALG_CHACHA20_POLY1305 ? 'C' : 'A',
+                   id->type == SESSION_ID_TYPE_SEQUENTIAL ? 'S' : 'R');
+}
+
+psa_key_id_t session_key_generate(const session_id_t *id,
+                                  psa_algorithm_t algorithm,
+                                  psa_key_id_t private_key,
+                                  const struct pubkey *pubkey,
+                                  psa_key_usage_t usage)
+{
+    psa_key_derivation_operation_t operation = PSA_KEY_DERIVATION_OPERATION_INIT;
+    psa_key_attributes_t key_attributes = PSA_KEY_ATTRIBUTES_INIT;
+    psa_key_id_t key = PSA_KEY_ID_NULL;
+    psa_status_t status;
+
+    if (pubkey->len == 0)
+    {
+        LOG_ERR("Missing server key");
+        return PSA_KEY_ID_NULL;
+    }
+
+    // ECDH + HKDF with SHA-256:
+    status = psa_key_derivation_setup(
+        &operation,
+        PSA_ALG_KEY_AGREEMENT(PSA_ALG_ECDH, PSA_ALG_HKDF(PSA_ALG_SHA_256)));
+    if (status != PSA_SUCCESS)
+    {
+        LOG_ERR("Couldn't set up key derivation: %d", status);
+        goto exit;
+    }
+
+    // Perform the ECDH key agreement, and feed the shared secret directly into the HKDF:
+    status = psa_key_derivation_key_agreement(&operation,
+                                              PSA_KEY_DERIVATION_INPUT_SECRET,
+                                              private_key,
+                                              pubkey->data,
+                                              pubkey->len);
+    if (status != PSA_SUCCESS)
+    {
+        LOG_ERR("Failed key agreement: %d", status);
+        goto exit;
+    }
+
+    uint8_t info[INFO_MAX_LEN];
+    ssize_t info_len = session_key_info_build(id, algorithm, info);
+    if (info_len < 0)
+    {
+        LOG_ERR("Failed session key build: %d", info_len);
+        goto exit;
+    }
+
+    status =
+        psa_key_derivation_input_bytes(&operation, PSA_KEY_DERIVATION_INPUT_INFO, info, info_len);
+    if (status != PSA_SUCCESS)
+    {
+        LOG_ERR("Failed info input: %d", status);
+        goto exit;
+    }
+
+    psa_set_key_lifetime(&key_attributes, PSA_KEY_LIFETIME_VOLATILE);
+    psa_set_key_type(&key_attributes, SESSION_KEY_TYPE(algorithm));
+    psa_set_key_bits(&key_attributes, SAEAD_KEY_SIZE * 8);
+    psa_set_key_algorithm(&key_attributes, algorithm);
+    psa_set_key_usage_flags(&key_attributes, usage);
+
+    status = psa_key_derivation_output_key(&key_attributes, &operation, &key);
+    if (status != PSA_SUCCESS)
+    {
+        LOG_ERR("Failed key derivation: %d", status);
+        goto exit;
+    }
+
+exit:
+    // Release the derivation context. Does not invalidate the key.
+    (void) psa_key_derivation_abort(&operation);
+    return key;
```

Comment:
We return the key to the caller here, but do we actually ever set it on the `session`? Is the caller responsible for setting the session key? I'm not necessarily opposed to exposing the PSA key ID outside of this module, but it does feel somewhat leaky given that we manage its lifecycle internally (e.g. ending session destroys key).

Reply:
We could pass in the whole session here to set the session key field, but I chose not to, as it makes it a bit hard to reason about which fields are expected to be populated and which will be set in the session object in this function. The session owners for uplink and downlink will be setting this, as well as the rest of the arguments in the session struct as soon as this passes

Reply:
Gotcha -- do you think it makes sense to remove the `algorithm` parameter from the `session_init` if that is the case?